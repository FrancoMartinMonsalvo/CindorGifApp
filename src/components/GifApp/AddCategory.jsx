import { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ onNewCategory }) => {

    const [inputValue, setInputValue] = useState([]);

    const onInputValue = ({ target }) => {
        setInputValue(target.value);
    }
    const onSubmit = (event) => {

        event.preventDefault();
        //el condicional de abajo lo que hara ser√° retornar si 
        //el largo del input es menor igual a 1, es decir, no tiene
        //caracteres, no manda el input
        if (inputValue.trim().length <= 1) return;

        setInputValue('');
        onNewCategory(inputValue.trim());

        //eso es para que una vez mande el input, se limpie el espacio

    }

    return (
        <form onSubmit={onSubmit} aria-label='form'>
            <input
                type="text"
                placeholder="Buscar gifs"
                value={inputValue}
                onChange={onInputValue}
            />
        </form>


    )
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired,

}
