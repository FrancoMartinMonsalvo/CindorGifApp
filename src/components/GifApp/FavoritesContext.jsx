import React, { createContext, useState, useEffect } from 'react';

export const FavoritesContext = createContext({
  favorites: [],
  toggleFavorite: () => {},
});

export const FavoritesProvider = ({ children }) => {
  const [favorites, setFavorites] = useState([]);

  useEffect(() => {
    // Cargar los favoritos desde localStorage
    const storedFavorites = JSON.parse(localStorage.getItem('favorites')) || [];

    // Validar los favoritos y filtrar los que no tienen id, title, o url
    const validFavorites = storedFavorites.filter(favGif => favGif.id && favGif.title && favGif.url);

    // Actualizar el estado y localStorage con los favoritos vÃ¡lidos
    setFavorites(validFavorites);
    localStorage.setItem('favorites', JSON.stringify(validFavorites));
  }, []);

  const toggleFavorite = (id) => {
    const updatedFavorites = favorites.includes(id)
      ? favorites.filter(favId => favId !== id)
      : [...favorites, id];

    setFavorites(updatedFavorites);
    localStorage.setItem('favorites', JSON.stringify(updatedFavorites));
  };

  return (
    <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>
      {children}
    </FavoritesContext.Provider>
  );
};
