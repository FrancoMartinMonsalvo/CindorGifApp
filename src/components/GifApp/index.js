// index.js (barril)
export * from './AddCategory';
export * from './GifExpertApp';
export * from './GifGrid';
export * from './GifGridItem';
export * from './FavoriteButton';
export * from './GifCollection';
export * from './FavoritesContext'; // Aseg√∫rate de que este archivo exporte FavoritesProvider
export * from './NavBar';
export * from './App';
