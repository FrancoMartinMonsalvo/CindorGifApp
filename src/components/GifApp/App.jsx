import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { GifExpertApp, NavBar } from '../GifApp'; // Ajusta las rutas seg√∫n tu estructura de archivos

export function App() {
  return (
    <BrowserRouter>
      <NavBar />
      <Routes>
        <Route path="/" element={<GifExpertApp />} />
        <Route path="/favoritos" element={<GifExpertApp favoritesOnly={true} />} />
      </Routes>
    </BrowserRouter>
  );
}

