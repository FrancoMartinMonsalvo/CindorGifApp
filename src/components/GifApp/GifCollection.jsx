// FavoriteCollection.jsx
import React, { useContext } from 'react';
import { FavoritesContext } from '../GifApp';

export const FavoriteCollection = () => {
  const { favorites } = useContext(FavoritesContext);

  return (
    <div>
      <h2>Mis favoritos</h2>
      <div className="card-grid">
        {favorites.map((gif) => (
          <div key={gif.id} className="card">
            <img src={gif.url} alt={gif.title} />
            <p>{gif.title}</p>
          </div>
        ))}
      </div>
    </div>
  );
};
