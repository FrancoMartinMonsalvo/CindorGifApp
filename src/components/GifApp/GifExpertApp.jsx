import { useState } from "react";
import { AddCategory, GifGrid, FavoriteCollection, FavoritesProvider } from '../GifApp';

export const GifExpertApp = ({ favoritesOnly = false }) => {
  const [categories, setCategories] = useState(['Sun Wukong']);

  const onAddCategory = (newCategory) => {
    if (categories.includes(newCategory)) return;
    setCategories([newCategory, ...categories]);
  };

  return (
    <FavoritesProvider>
      {/* Input de b√∫squeda al principio */}
      <AddCategory onNewCategory={onAddCategory} />

      {favoritesOnly ? (
        <FavoriteCollection />
      ) : (
        <>
          {categories.map((category) => (
            <GifGrid key={category} category={category} />
          ))}
        </>
      )}
    </FavoritesProvider>
  );
};
