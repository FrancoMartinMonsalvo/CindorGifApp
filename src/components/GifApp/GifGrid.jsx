import { GifGridItem } from '../GifApp';
import { useFetchGifs } from "../../hooks/SearchGif/useFetchGifs";
import PropTypes from 'prop-types';


export const GifGrid = ({ category }) => {
  const {images, isLoading} = useFetchGifs(category);

  return (
    <>
      <h3> {category} </h3>
      { isLoading && (<h2> Cargando </h2>)}
      <div className="card-grid">
        {
          images?.map( ( img  ) => (
            <GifGridItem key={img.id}
                         { ... img }
            />  
          ))
        }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,

}

