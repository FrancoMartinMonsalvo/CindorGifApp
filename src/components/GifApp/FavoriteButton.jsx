import React, { useContext } from 'react';
import { FavoritesContext } from './FavoritesContext';

export const FavoriteButton = ({ id, title, url }) => {
  const { favorites, toggleFavorite } = useContext(FavoritesContext);

  const isFavorite = favorites.some(favGif => favGif.id === id);

  return (
    <button className="fav-item" onClick={() => toggleFavorite({ id, title, url })}>
      {isFavorite ? '🖤' : '🤍'}
    </button>
  );
};
