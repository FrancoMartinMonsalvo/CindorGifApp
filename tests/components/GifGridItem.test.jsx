import { render, screen } from "@testing-library/react";
import { GifGridItem } from "../../src/components/GifApp";

describe('Pruebas en GifGridItem', () => { 

    const title = 'Baki';
    const url = 'http://localhost/https//uBaki.com/baki.jpg';

    test('Debe de hacer match con el snapshot', ()=> {

        

        const { container } = render( <GifGridItem title={title} url={url} /> );
        expect( container ).toMatchSnapshot();
    
        
    });

    test('Debe de mostrar la imagen con el URL y el ALT indicado', () => {
        
        render( <GifGridItem title={title} url={url} /> );
        //screen.debug();
        //expect( screen.getByRole('img').src).toStrictEqual(url);
        //expect( screen.getByRole('img').alt).toStrictEqual(title);
        const {src,alt} = screen.getByRole('img'); //Es una manera mas elegante de hacer lo de arriba, desestructurando 
        expect(src).toBe(url);
        expect(alt).toBe(title);
    });
    test('debe de mostrar el titulo en el componente', () => {
        
        render( <GifGridItem title={title} url={url} /> );
        expect(screen.getByText(title)).toBeTruthy();


    });
    
    
    
    

});
 