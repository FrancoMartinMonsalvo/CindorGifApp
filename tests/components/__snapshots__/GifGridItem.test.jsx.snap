// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGridItem Debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Baki"
      src="http://localhost/https//uBaki.com/baki.jpg"
    />
    <p>
      Baki
    </p>
    <button
      class="fav-item"
    >
      ü§ç
    </button>
  </div>
</div>
`;
